// virtual-alexa/lib/src/dialog/DialogIntent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var DialogIntent = (function () {
    function DialogIntent() {
    }
    DialogIntent.fromJSON = function (interactionModel, json) {
        var intent = new DialogIntent();
        Object.assign(intent, json);
        intent.slots = [];
        for (var _i = 0, _a = json.slots; _i < _a.length; _i++) {
            var slot = _a[_i];
            intent.addSlot(DialogSlot.fromJSON(intent, slot));
        }
        return intent;
    };
    DialogIntent.prototype.addSlot = function (slot) {
        this.slots.push(slot);
    };
    return DialogIntent;
}());
exports.DialogIntent = DialogIntent;
var DialogSlot = (function () {
    function DialogSlot(dialogIntent) {
        this.dialogIntent = dialogIntent;
    }
    DialogSlot.fromJSON = function (dialogIntent, json) {
        var slot = new DialogSlot(dialogIntent);
        Object.assign(slot, json);
        return slot;
    };
    DialogSlot.prototype.elicitationPrompt = function () {
        var id = this.prompts.elicitation;
        return this.dialogIntent.interactionModel.prompt(id);
    };
    DialogSlot.prototype.confirmationPrompt = function () {
        var id = this.prompts.confirmation;
        return this.dialogIntent.interactionModel.prompt(id);
    };
    return DialogSlot;
}());
exports.DialogSlot = DialogSlot;
