// bespoken-tools/lib/alexa/alexa.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var audio_player_1 = require("./audio-player");
var logging_helper_1 = require("../core/logging-helper");
var service_request_1 = require("./service-request");
var request = require("request");
var alexa_context_1 = require("./alexa-context");
var events_1 = require("events");
var Logger = "ALEXA";
var AlexaEvent;
(function (AlexaEvent) {
    AlexaEvent[AlexaEvent["SessionEnded"] = 0] = "SessionEnded";
    AlexaEvent[AlexaEvent["SkillError"] = 1] = "SkillError";
    AlexaEvent[AlexaEvent["SkillResponse"] = 2] = "SkillResponse";
})(AlexaEvent = exports.AlexaEvent || (exports.AlexaEvent = {}));
var Alexa = (function () {
    function Alexa() {
        this._actionQueue = new ActionQueue();
        this._context = null;
        this._emitter = null;
        this._emitter = new events_1.EventEmitter();
    }
    Alexa.prototype.startSession = function (skillURL, model, audioEnabled, applicationID) {
        var audioPlayer = null;
        if (audioEnabled) {
            audioPlayer = new audio_player_1.AudioPlayer(this);
        }
        this._context = new alexa_context_1.AlexaContext(skillURL, model, audioPlayer, applicationID);
        this._context.newSession();
        return this;
    };
    Alexa.prototype.context = function () {
        return this._context;
    };
    Alexa.prototype.interactionModel = function () {
        return this.context().interactionModel();
    };
    Alexa.prototype.spoken = function (utterance, callback) {
        var intent = this.interactionModel().sampleUtterances.intentForUtterance(utterance);
        if (intent === null) {
            var defaultUtterance = this.interactionModel().sampleUtterances.defaultUtterance();
            intent = this.interactionModel().sampleUtterances.intentForUtterance(defaultUtterance);
            logging_helper_1.LoggingHelper.warn(Logger, "No intentName matches utterance: " + utterance + ". Using fallback utterance: " + intent.utterance);
        }
        this.callSkillWithIntent(intent.intentName, intent.toJSON(), callback);
    };
    Alexa.prototype.launched = function (callback) {
        var serviceRequest = new service_request_1.ServiceRequest(this._context);
        serviceRequest.launchRequest();
        this.callSkill(serviceRequest, callback);
    };
    Alexa.prototype.sessionEnded = function (sessionEndedReason, errorData, callback) {
        if (sessionEndedReason === service_request_1.SessionEndedReason.ERROR) {
            logging_helper_1.LoggingHelper.error(Logger, "SessionEndedRequest:\n" + JSON.stringify(errorData, null, 2));
        }
        var serviceRequest = new service_request_1.ServiceRequest(this._context);
        serviceRequest.sessionEndedRequest(sessionEndedReason, errorData);
        this.callSkill(serviceRequest, callback);
        this.context().endSession();
    };
    Alexa.prototype.intended = function (intentName, slots, callback) {
        this.callSkillWithIntent(intentName, slots, callback);
    };
    Alexa.prototype.callSkillWithIntent = function (intentName, slots, callback) {
        var self = this;
        try {
            if (this._context.audioPlayerEnabled() && this._context.audioPlayer().isPlaying()) {
                this._context.audioPlayer().suspend();
            }
            var serviceRequest = new service_request_1.ServiceRequest(this._context).intentRequest(intentName);
            if (slots !== undefined && slots !== null) {
                for (var _i = 0, _a = Object.keys(slots); _i < _a.length; _i++) {
                    var slotName = _a[_i];
                    serviceRequest.withSlot(slotName, slots[slotName]);
                }
            }
            this.callSkill(serviceRequest, function (error, responseJSON, requestJSON) {
                if (callback !== undefined && callback !== null) {
                    callback(error, responseJSON, requestJSON);
                }
                if (self._context.audioPlayerEnabled() && self._context.audioPlayer().suspended()) {
                    self._context.audioPlayer().resume();
                }
            });
        }
        catch (e) {
            if (callback !== undefined && callback !== null) {
                callback(e, null, null);
            }
        }
    };
    Alexa.prototype.callSkill = function (serviceRequest, callback) {
        var self = this;
        this.sequence(function (done) {
            self.callSkillImpl(serviceRequest, callback, done);
        });
    };
    Alexa.prototype.sequence = function (action) {
        this._actionQueue.enqueue(action);
    };
    Alexa.prototype.callSkillImpl = function (serviceRequest, callback, done) {
        var self = this;
        if (serviceRequest.requiresSession() && !this.context().activeSession()) {
            this.context().newSession();
        }
        var requestJSON = serviceRequest.toJSON();
        logging_helper_1.LoggingHelper.debug(Logger, "CALLING: " + requestJSON.request.type);
        var responseHandler = function (error, response, body) {
            if (self.context().activeSession()) {
                self.context().session().used();
                if (!error) {
                    if (body.response !== undefined && body.response.shouldEndSession) {
                        self.context().endSession();
                    }
                    else {
                        self.context().session().updateAttributes(body.sessionAttributes);
                    }
                }
            }
            if (error) {
                if (callback !== undefined && callback !== null) {
                    callback(error, null, null);
                }
            }
            else {
                if (body.response !== undefined && body.response.directives !== undefined) {
                    self._context.audioPlayer().directivesReceived(body.response.directives);
                }
                if (callback !== undefined && callback !== null) {
                    callback(null, body, requestJSON);
                }
                self._emitter.emit(AlexaEvent[AlexaEvent.SkillResponse], body, requestJSON);
            }
            done();
        };
        this.post({
            url: this.context().skillURL(),
            method: "POST",
            json: requestJSON,
        }, responseHandler);
    };
    Alexa.prototype.post = function (options, responseHandler) {
        request.post(options, responseHandler);
    };
    Alexa.prototype.on = function (event, callback) {
        this._emitter.on(AlexaEvent[event], callback);
    };
    Alexa.prototype.once = function (event, callback) {
        this._emitter.once(AlexaEvent[event], callback);
    };
    Alexa.prototype.stop = function (onStop) {
        this._actionQueue.stop(function () {
            onStop();
        });
    };
    return Alexa;
}());
exports.Alexa = Alexa;
var ActionQueue = (function () {
    function ActionQueue() {
        this._queue = [];
        this._stop = false;
    }
    ActionQueue.prototype.enqueue = function (action) {
        this._queue.push(action);
        if (this._queue.length === 1) {
            this.next();
        }
    };
    ActionQueue.prototype.processing = function () {
        return (this._queue.length > 0);
    };
    ActionQueue.prototype.next = function () {
        var self = this;
        if (this._queue.length === 0) {
            return;
        }
        var action = this._queue[0];
        action(function () {
            self._queue = self._queue.slice(1);
            if (self._stop) {
                self._stopCallback();
            }
            else {
                self.next();
            }
        });
    };
    ActionQueue.prototype.stop = function (onStop) {
        this._stop = true;
        if (this.processing()) {
            this._stopCallback = onStop;
        }
        else {
            onStop();
        }
    };
    return ActionQueue;
}());
exports.ActionQueue = ActionQueue;
