// bespoken-tools/lib/alexa/alexa-context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var alexa_session_1 = require("./alexa-session");
var uuid = require("uuid");
var AlexaContext = (function () {
    function AlexaContext(_skillURL, _interactionModel, _audioPlayer, _applicationID) {
        this._skillURL = _skillURL;
        this._interactionModel = _interactionModel;
        this._audioPlayer = _audioPlayer;
        this._applicationID = _applicationID;
        this._accessToken = null;
    }
    AlexaContext.prototype.applicationID = function () {
        if (this._applicationID === undefined || this._applicationID === null) {
            this._applicationID = "amzn1.echo-sdk-ams.app." + uuid.v4();
        }
        return this._applicationID;
    };
    AlexaContext.prototype.skillURL = function () {
        return this._skillURL;
    };
    AlexaContext.prototype.interactionModel = function () {
        return this._interactionModel;
    };
    AlexaContext.prototype.userID = function () {
        if (this._userID === undefined || this._userID === null) {
            this._userID = "amzn1.ask.account." + uuid.v4();
        }
        return this._userID;
    };
    AlexaContext.prototype.setUserID = function (userID) {
        this._userID = userID;
    };
    AlexaContext.prototype.accessToken = function () {
        return this._accessToken;
    };
    AlexaContext.prototype.setAccessToken = function (token) {
        this._accessToken = token;
    };
    AlexaContext.prototype.audioPlayer = function () {
        return this._audioPlayer;
    };
    AlexaContext.prototype.audioPlayerEnabled = function () {
        return this._audioPlayer !== null;
    };
    AlexaContext.prototype.newSession = function () {
        this._session = new alexa_session_1.AlexaSession();
    };
    AlexaContext.prototype.session = function () {
        return this._session;
    };
    AlexaContext.prototype.endSession = function () {
        this._session = null;
    };
    AlexaContext.prototype.activeSession = function () {
        return this._session !== null;
    };
    return AlexaContext;
}());
exports.AlexaContext = AlexaContext;
