// bespoken-tools/lib/core/webhook-request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var querystring = require("querystring");
var WebhookRequest = (function () {
    function WebhookRequest(sourceSocket) {
        this.sourceSocket = sourceSocket;
        this.queryParameters = {};
        this.rawContents = new Buffer("");
        this.rawBody = new Buffer("");
        this.body = "";
        this.requestID = new Date().getTime();
        if (this.sourceSocket === undefined) {
            this.sourceSocket = null;
        }
    }
    WebhookRequest.fromBuffer = function (sourceSocket, payload, id) {
        var webhookRequest = new WebhookRequest(sourceSocket);
        webhookRequest.append(payload);
        webhookRequest.requestID = id;
        return webhookRequest;
    };
    WebhookRequest.prototype.append = function (data) {
        this.rawContents = Buffer.concat([this.rawContents, data]);
        if (this.headers == null) {
            this.headers = {};
            var endIndex = this.rawContents.indexOf("\r\n\r\n");
            if (endIndex !== -1) {
                this.parseHeaders(this.rawContents.slice(0, endIndex).toString());
                if (endIndex + 4 < this.rawContents.length) {
                    var bodyPart = this.rawContents.slice((endIndex + 4));
                    this.appendBody(bodyPart);
                }
            }
        }
        else {
            this.appendBody(data);
        }
    };
    WebhookRequest.prototype.appendBody = function (bodyPart) {
        this.rawBody = Buffer.concat([this.rawBody, bodyPart]);
        this.body += bodyPart.toString();
    };
    WebhookRequest.prototype.done = function () {
        if (this.method === "GET") {
            return true;
        }
        return (this.rawBody.length === this.contentLength());
    };
    WebhookRequest.prototype.contentLength = function () {
        var contentLength = -1;
        if (this.headers != null) {
            var contentLengthString = this.headers["content-length"];
            contentLength = parseInt(contentLengthString);
        }
        return contentLength;
    };
    WebhookRequest.prototype.isPing = function () {
        return (this.uri.indexOf("/ping") !== -1);
    };
    WebhookRequest.prototype.parseHeaders = function (headersString) {
        var lines = headersString.split("\n");
        var requestLine = lines[0];
        var requestLineParts = requestLine.split(" ");
        this.method = requestLineParts[0];
        this.uri = requestLineParts[1];
        if (this.uri.indexOf("?") >= 0) {
            var qs = this.uri.replace(/^.*\?/, "");
            this.queryParameters = querystring.parse(qs);
        }
        for (var i = 1; i < lines.length; i++) {
            var headerLine = lines[i];
            var headerParts = headerLine.split(":");
            var key = headerParts[0].toLowerCase();
            this.headers[key] = headerParts[1].trim();
        }
    };
    WebhookRequest.prototype.nodeID = function () {
        var nodeValue = null;
        if ("node-id" in this.queryParameters) {
            if (typeof this.queryParameters["node-id"] === "string") {
                nodeValue = this.queryParameters["node-id"];
            }
            else {
                throw new Error("Only one node-id should be present in the query");
            }
        }
        return nodeValue;
    };
    WebhookRequest.prototype.removeBespokenQueries = function (httpLine) {
        return httpLine.replace("&node-id=" + this.nodeID(), "")
            .replace("node-id=" + this.nodeID(), "")
            .replace("&bespoken-key=" + this.nodeID(), "")
            .replace("bespoken-key=" + this.nodeID(), "")
            .replace("? HTTP", " HTTP");
    };
    WebhookRequest.prototype.requestWithoutBespokenData = function () {
        var firstLineBreak = this.rawContents.indexOf("\n");
        var httpLine = this.rawContents.slice(0, firstLineBreak).toString();
        return Buffer.concat([Buffer.from(this.removeBespokenQueries(httpLine)),
            this.rawContents.slice(firstLineBreak)]);
    };
    WebhookRequest.prototype.isJSON = function () {
        try {
            JSON.parse(this.body);
            return true;
        }
        catch (error) {
            return false;
        }
    };
    WebhookRequest.prototype.toString = function () {
        return this.method + " " + this.uri;
    };
    WebhookRequest.prototype.id = function () {
        return this.requestID;
    };
    return WebhookRequest;
}());
exports.WebhookRequest = WebhookRequest;
