// virtual-alexa/lib/src/external/AddressAPI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var nock = require("nock");
var AddressAPI = (function () {
    function AddressAPI(context) {
        this.context = context;
        this.reset();
    }
    AddressAPI.prototype.returnsFullAddress = function (address) {
        this.configure(200, "/settings/address", address);
        var countryAndPostalCode = {
            countryCode: address.countryCode,
            postalCode: address.postalCode,
        };
        this.configure(200, "/settings/address/countryAndPostalCode", countryAndPostalCode);
    };
    AddressAPI.prototype.returnsCountryAndPostalCode = function (address) {
        this.configure(200, "/settings/address/countryAndPostalCode", address);
        this.configure(403, "/settings/address");
    };
    AddressAPI.prototype.insufficientPermissions = function () {
        this.configure(403, "/settings/address/countryAndPostalCode");
        this.configure(403, "/settings/address");
    };
    AddressAPI.prototype.reset = function () {
        if (AddressAPI.addressScope) {
            AddressAPI.addressScope.persist(false);
        }
        if (AddressAPI.postalScope) {
            AddressAPI.postalScope.persist(false);
        }
    };
    AddressAPI.prototype.configure = function (responseCode, pathEnd, payload) {
        var _this = this;
        if (!nock.isActive()) {
            nock.activate();
        }
        if (payload) {
            this.context.device().generatedID();
        }
        var baseURL = this.context.apiEndpoint();
        var scope = nock(baseURL)
            .persist()
            .get(function (path) {
            return path === ("/v1/devices/" + _this.context.device().id() + pathEnd);
        })
            .query(true)
            .reply(responseCode, JSON.stringify(payload, null, 2));
        if (pathEnd.endsWith("countryAndPostalCode")) {
            AddressAPI.postalScope = scope;
        }
        else {
            AddressAPI.addressScope = scope;
        }
    };
    return AddressAPI;
}());
exports.AddressAPI = AddressAPI;
