// virtual-alexa/lib/src/external/DynamoDB.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var nock = require("nock");
var AWS = require("aws-sdk");
var DynamoDB = (function () {
    function DynamoDB() {
        this.records = [];
        this.region = "us-east-1";
    }
    DynamoDB.prototype.mock = function () {
        this.setEnv();
        if (!nock.isActive()) {
            nock.activate();
        }
        this.mockPut();
        this.mockGet();
        this.mockCreate();
    };
    DynamoDB.prototype.reset = function () {
        this.records = [];
        if (DynamoDB.getScope) {
            DynamoDB.getScope.persist(false);
        }
        if (DynamoDB.putScope) {
            DynamoDB.putScope.persist(false);
        }
        if (DynamoDB.createScope) {
            DynamoDB.createScope.persist(false);
        }
    };
    DynamoDB.prototype.setEnv = function () {
        if (process.env.AWS_REGION) {
            this.region = process.env.AWS_REGION;
        }
        this.setDefault("AWS_REGION", "us-east-1");
        this.setDefault("AWS_ACCESS_KEY_ID", "123456789");
        this.setDefault("AWS_SECRET_ACCESS_KEY", "123456789");
    };
    DynamoDB.prototype.setDefault = function (property, value) {
        process.env[property] = process.env[property] ? process.env[property] : value;
    };
    DynamoDB.prototype.fetchImpl = function (table, key) {
        for (var i = this.records.length - 1; i >= 0; i--) {
            var record = this.records[i];
            if (record.TableName !== table) {
                continue;
            }
            var o = this.simplifyRecord(record.Item);
            var match = true;
            for (var _i = 0, _a = Object.keys(key); _i < _a.length; _i++) {
                var keyPart = _a[_i];
                var keyPartValue = key[keyPart];
                if (o[keyPart] && o[keyPart] === keyPartValue) {
                    continue;
                }
                match = false;
                break;
            }
            if (match) {
                return record;
            }
        }
        return undefined;
    };
    DynamoDB.prototype.mockPut = function () {
        var _this = this;
        DynamoDB.putScope = nock(this.baseURL())
            .matchHeader("x-amz-target", function (value) {
            return value.endsWith("PutItem");
        })
            .persist()
            .post("/", function (body) {
            _this.records.push(body);
            return true;
        })
            .query(true)
            .reply(200, JSON.stringify({}));
    };
    DynamoDB.prototype.mockGet = function () {
        var _this = this;
        DynamoDB.getScope = nock(this.baseURL())
            .matchHeader("x-amz-target", function (value) {
            return value.endsWith("GetItem");
        })
            .persist()
            .post("/", function (body) {
            return true;
        })
            .query(true)
            .reply(200, function (uri, requestBody) {
            var requestObject = JSON.parse(requestBody);
            var key = requestObject.Key;
            var keySimple = _this.simplifyRecord(key);
            var record = _this.fetchImpl(requestObject.TableName, keySimple);
            if (!record) {
                record = {};
            }
            return record;
        });
    };
    DynamoDB.prototype.mockCreate = function () {
        DynamoDB.createScope = nock(this.baseURL())
            .matchHeader("x-amz-target", function (value) {
            return value.endsWith("CreateTable");
        })
            .persist()
            .post("/", function (body) {
            return true;
        })
            .query(true)
            .reply(200, function (uri, requestBody) {
            var bodyJSON = JSON.parse(requestBody);
            var response = {
                TableDescription: bodyJSON,
            };
            response.TableDescription.TableStatus = "CREATING";
            return response;
        });
    };
    DynamoDB.prototype.baseURL = function () {
        return "https://dynamodb." + this.region + ".amazonaws.com:443";
    };
    DynamoDB.prototype.simplifyRecord = function (dynamoObject) {
        return AWS.DynamoDB.Converter.unmarshall(dynamoObject);
    };
    return DynamoDB;
}());
exports.DynamoDB = DynamoDB;
