// bespoken-tools/lib/server/bespoke-server.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var node_manager_1 = require("./node-manager");
var webhook_manager_1 = require("./webhook-manager");
var http_helper_1 = require("../core/http-helper");
var global_1 = require("../core/global");
var logging_helper_1 = require("../core/logging-helper");
var bst_statistics_1 = require("../statistics/bst-statistics");
var Logger = "BSPKD";
var BespokeServer = (function () {
    function BespokeServer(webhookPort, nodePort) {
        this.webhookPort = webhookPort;
        this.nodePort = nodePort;
    }
    BespokeServer.prototype.start = function (started) {
        bst_statistics_1.BstStatistics.instance().start();
        var self = this;
        console.error("AWS_KEY: " + process.env["AWS_ACCESS_KEY_ID"]);
        var count = 0;
        var callbackCounter = function () {
            count++;
            if (count === 2) {
                if (started !== undefined && started !== null) {
                    started();
                }
            }
        };
        this.nodeManager = new node_manager_1.NodeManager(this.nodePort);
        this.nodeManager.start(callbackCounter);
        this.webhookManager = new webhook_manager_1.WebhookManager(this.webhookPort);
        this.webhookManager.start(callbackCounter);
        this.webhookManager.onWebhookReceived = function (webhookRequest) {
            if (webhookRequest.isPing()) {
                http_helper_1.HTTPHelper.respond(webhookRequest.sourceSocket, 200, "bst-server-" + global_1.Global.version());
            }
            else {
                try {
                    webhookRequest.nodeID();
                }
                catch (error) {
                    http_helper_1.HTTPHelper.respond(webhookRequest.sourceSocket, 400, error.message);
                    bst_statistics_1.BstStatistics.instance().record(bst_statistics_1.BstCommand.proxy, bst_statistics_1.BstEvent.dropped);
                    return;
                }
                if (webhookRequest.nodeID() === null) {
                    logging_helper_1.LoggingHelper.error(Logger, "No node specified: " + webhookRequest.uri);
                    http_helper_1.HTTPHelper.respond(webhookRequest.sourceSocket, 400, "No node specified. Must be included with the querystring as node-id.");
                }
                else {
                    var node = self.nodeManager.node(webhookRequest.nodeID());
                    if (node == null) {
                        logging_helper_1.LoggingHelper.error(Logger, "Node is not active: " + webhookRequest.nodeID());
                        http_helper_1.HTTPHelper.respond(webhookRequest.sourceSocket, 404, "Node is not active: " + webhookRequest.nodeID());
                        bst_statistics_1.BstStatistics.instance().record(bst_statistics_1.BstCommand.proxy, bst_statistics_1.BstEvent.dropped, webhookRequest.nodeID());
                    }
                    else {
                        logging_helper_1.LoggingHelper.info(Logger, "Forwarded: " + webhookRequest.nodeID());
                        node.forward(webhookRequest);
                        bst_statistics_1.BstStatistics.instance().record(bst_statistics_1.BstCommand.proxy, bst_statistics_1.BstEvent.forwarded, node.id);
                    }
                }
            }
        };
        this.uncaughtExceptionHandler = function (error) {
            console.error("UncaughtException: " + error.stack);
        };
        process.on("uncaughtException", this.uncaughtExceptionHandler);
    };
    BespokeServer.prototype.stop = function (callback) {
        bst_statistics_1.BstStatistics.instance().stop();
        var count = 0;
        var callbackFunction = function () {
            count++;
            if (count === 2) {
                callback();
            }
        };
        process.removeListener("uncaughtException", this.uncaughtExceptionHandler);
        this.nodeManager.stop(callbackFunction);
        this.webhookManager.stop(callbackFunction);
    };
    return BespokeServer;
}());
exports.BespokeServer = BespokeServer;
