// bespoken-tools/lib/client/function-server.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var logging_helper_1 = require("../core/logging-helper");
var module_manager_1 = require("./module-manager");
var bodyParser = require("body-parser");
var express = require("express");
var Logger = "BST-FUNCTION";
var FunctionServer = (function () {
    function FunctionServer(file, functionName, port, verbose) {
        this.file = file;
        this.functionName = functionName;
        this.port = port;
        this.verbose = verbose;
        this.requests = [];
        this.server = null;
    }
    FunctionServer.prototype.start = function (callback) {
        var self = this;
        this.moduleManager = new module_manager_1.ModuleManager(process.cwd());
        this.moduleManager.start();
        var app = express();
        app.use(bodyParser.json());
        app.use(function (request, response) {
            self.requests.push(request);
            self.invoke(request, response);
        });
        this.server = app.listen(this.port, function () {
            logging_helper_1.LoggingHelper.info(Logger, "CloudFunctionServer started on port: " + self.port);
            if (callback !== undefined && callback !== null) {
                callback();
            }
        });
    };
    FunctionServer.prototype.stop = function (onStop) {
        this.moduleManager.stop();
        var request = null;
        for (var _i = 0, _a = this.requests; _i < _a.length; _i++) {
            request = _a[_i];
            try {
                request.socket.end();
            }
            catch (e) {
            }
        }
        this.server.close(function () {
            if (onStop !== undefined && onStop !== null) {
                onStop();
            }
        });
    };
    FunctionServer.prototype.invoke = function (request, response) {
        var path = this.file;
        logging_helper_1.LoggingHelper.debug(Logger, "Invoking Function: " + this.file);
        var cloudFunction = this.moduleManager.module(path);
        if (!(this.functionName in cloudFunction)) {
            var message = "Function: " + this.functionName
                + " does not exist or has not been exported from module: " + this.file;
            logging_helper_1.LoggingHelper.error(Logger, message);
            response.status(500).send(message);
            return;
        }
        try {
            cloudFunction[this.functionName].call(cloudFunction, request, response);
        }
        catch (e) {
            if (e.stack) {
                console.error(e.stack);
            }
            else {
                console.error(e);
            }
            var message = "Unhandled Exception from Cloud Function: " + e;
            response.status(500).send(message);
        }
    };
    return FunctionServer;
}());
exports.FunctionServer = FunctionServer;
