// bespoken-tools/lib/client/module-manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("fs");
var logging_helper_1 = require("../core/logging-helper");
var node_util_1 = require("../core/node-util");
var path = require("path");
var Logger = "MODULE-MGR";
var ModuleManager = (function () {
    function ModuleManager(directory) {
        this.directory = directory;
        this.onDirty = null;
        this.dirty = false;
        this.modules = {};
        this.watcher = null;
        if (!path.isAbsolute(directory)) {
            this.directory = path.join(process.cwd(), this.directory);
        }
    }
    ModuleManager.prototype.start = function () {
        var self = this;
        var watchOptions = { "persistent": false, "recursive": true };
        this.watcher = fs.watch(this.directory, watchOptions, function (event, filename) {
            var exclude = false;
            if (filename.indexOf("node_modules") !== -1) {
                exclude = true;
            }
            else if (filename.endsWith("___")) {
                exclude = true;
            }
            else if (filename.startsWith(".")) {
                exclude = true;
            }
            if (!exclude) {
                logging_helper_1.LoggingHelper.info(Logger, "FS.Watch Event: " + event + ". File: " + filename + ". Reloading.");
                self.dirty = true;
                if (self.onDirty !== undefined && self.onDirty !== null) {
                    self.onDirty(filename);
                }
            }
        });
    };
    ModuleManager.prototype.module = function (filePath) {
        var fullPath = path.join(this.directory, filePath);
        var module = null;
        if (fullPath in this.modules && !this.dirty) {
            module = this.modules[fullPath];
        }
        else {
            module = node_util_1.NodeUtil.load(fullPath);
            this.modules[fullPath] = module;
            this.dirty = false;
        }
        return module;
    };
    ModuleManager.prototype.stop = function () {
        this.watcher.close();
    };
    return ModuleManager;
}());
exports.ModuleManager = ModuleManager;
