// virtual-alexa/lib/src/core/Device.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var uuid = require("uuid");
var Device = (function () {
    function Device(id) {
        this._supportedInterfaces = {};
        this._id = id;
        this.audioPlayerSupported(true);
    }
    Device.prototype.id = function () {
        return this._id;
    };
    Device.prototype.generatedID = function () {
        if (this._id) {
            return;
        }
        this._id = "virtualAlexa.deviceID." + uuid.v4();
    };
    Device.prototype.setID = function (id) {
        this._id = id;
    };
    Device.prototype.audioPlayerSupported = function (value) {
        return this.supportedInterface("AudioPlayer", value);
    };
    Device.prototype.displaySupported = function (value) {
        return this.supportedInterface("Display", value);
    };
    Device.prototype.videoAppSupported = function (value) {
        return this.supportedInterface("VideoApp", value);
    };
    Device.prototype.supportedInterfaces = function () {
        return this._supportedInterfaces;
    };
    Device.prototype.supportedInterface = function (name, value) {
        if (value !== undefined) {
            if (value === true) {
                this._supportedInterfaces[name] = {};
            }
            else {
                delete this._supportedInterfaces[name];
            }
        }
        return this._supportedInterfaces[name] !== undefined;
    };
    return Device;
}());
exports.Device = Device;
