// virtual-alexa/lib/src/core/SkillContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var uuid = require("uuid");
var DialogManager_1 = require("../dialog/DialogManager");
var Device_1 = require("./Device");
var SkillSession_1 = require("./SkillSession");
var User_1 = require("./User");
var SkillContext = (function () {
    function SkillContext(interactionModel, audioPlayer, _locale, _applicationID) {
        this._locale = _locale;
        this._applicationID = _applicationID;
        this._accessToken = null;
        this._apiAccessToken = "virtualAlexa.accessToken." + uuid.v4();
        this._apiEndpoint = "https://api.amazonalexa.com";
        this._audioPlayer = audioPlayer;
        this._interactionModel = interactionModel;
        this._dialogManager = new DialogManager_1.DialogManager(this);
        this._device = new Device_1.Device();
        this._user = new User_1.User();
    }
    SkillContext.prototype.apiAccessToken = function () {
        return this._apiAccessToken;
    };
    SkillContext.prototype.apiEndpoint = function () {
        return this._apiEndpoint;
    };
    SkillContext.prototype.applicationID = function () {
        if (this._applicationID === undefined || this._applicationID === null) {
            this._applicationID = "amzn1.echo-sdk-ams.app." + uuid.v4();
        }
        return this._applicationID;
    };
    SkillContext.prototype.device = function () {
        return this._device;
    };
    SkillContext.prototype.interactionModel = function () {
        return this._interactionModel;
    };
    SkillContext.prototype.user = function () {
        return this._user;
    };
    SkillContext.prototype.accessToken = function () {
        return this._accessToken;
    };
    SkillContext.prototype.setAccessToken = function (token) {
        this._accessToken = token;
    };
    SkillContext.prototype.locale = function () {
        return this._locale ? this._locale : "en-US";
    };
    SkillContext.prototype.dialogManager = function () {
        return this._dialogManager;
    };
    SkillContext.prototype.audioPlayer = function () {
        return this._audioPlayer;
    };
    SkillContext.prototype.newSession = function () {
        this._session = new SkillSession_1.SkillSession();
    };
    SkillContext.prototype.session = function () {
        return this._session;
    };
    SkillContext.prototype.endSession = function () {
        this.dialogManager().reset();
        this._session = undefined;
    };
    SkillContext.prototype.activeSession = function () {
        return this._session !== undefined;
    };
    return SkillContext;
}());
exports.SkillContext = SkillContext;
